src/lib/OpenEXR/ImfContextInit.cpp

```diff
@@ -0,0 +1,213 @@
+        {
+            // bah, there could be two reasons for this, one is a
+            // legitimate error, the other is a read past the end of file
+            // (i.e. the core library tries to read a 4k block when
+            // parsing the header), let's let the core deal with that and
+            // clear errors
+            ih->_stream->clear ();
+            //error_cb (
+            //    ctxt,
+            //    EXR_ERR_READ_IO,
+            //    "Unable to read requested bytes: %" PRIu64,
+            //    sz);
+        }
+        nread = s->tellg () - nread;
+    }
+    catch (...)
+    {
+        nread = -1;
+    }
+    return nread;
+}
+
+static void
+istream_destroy (exr_const_context_t ctxt, void* userdata, int failed)
+{
+    istream_holder* ih = static_cast<istream_holder*> (userdata);
+    delete ih;
+}
+
+ContextInitializer&
+ContextInitializer::setInputStream (IStream* istr)
+{
+    _initializer.user_data  = new istream_holder{istr};
+    _initializer.read_fn    = istream_read;
+    // TODO: add query to io streams to ask size if possible (such
+    // that ifstream can return size, others can return -1)
+    _initializer.size_fn    = nullptr; //istream_guess_size;
+    _initializer.write_fn   = nullptr;
+    _initializer.destroy_fn = istream_destroy;
+    _ctxt_type              = ContextFileType::READ;
+    _prov_stream            = istr;
+    return *this;
+}
+
+struct ostream_holder
+{
+    ostream_holder (OStream* s) : _stream (s)
+    {
+        if (_stream) _cur_offset = _stream->tellp ();
+    }
+
+#if ILMTHREAD_THREADING_ENABLED
+    std::mutex _mx;
+#endif
+    uint64_t _cur_offset = 0;
+    OStream* _stream;
+};
+
+static int64_t
+ostream_write (
+    exr_const_context_t         ctxt,
+    void*                       userdata,
+    const void*                 buffer,
+    uint64_t                    sz,
+    uint64_t                    offset,
+    exr_stream_error_func_ptr_t error_cb)
+{
+    ostream_holder* oh = static_cast<ostream_holder*> (userdata);
+
+    if (sz > INT_MAX)
+    {
+        error_cb (
+            ctxt,
+            EXR_ERR_READ_IO,
+            "Stream interface request to write block too large");
+        return -1;
+    }
+
+#if ILMTHREAD_THREADING_ENABLED
+    std::lock_guard<std::mutex> lk{oh->_mx};
+#endif
+
+    if (offset != oh->_cur_offset)
+    {
+        oh->_stream->seekp (offset);
+        oh->_cur_offset = oh->_stream->tellp ();
+        if (offset != oh->_cur_offset)
+        {
+            error_cb (
+                ctxt,
+                EXR_ERR_READ_IO,
+                "Unable to seek to desired offset %" PRIu64,
+                offset);
+            return -1;
+        }
+    }
+
+    int64_t nwrite = oh->_cur_offset;
+    try
+    {
+        oh->_stream->write (
+            static_cast<const char*> (buffer), static_cast<int> (sz));
+        oh->_cur_offset = oh->_stream->tellp ();
+        nwrite          = oh->_cur_offset - nwrite;
+    }
+    catch (...)
```

Comment:
do we want to be more specific in all of these generic catches? What I'm wondering about is that the throws might have information in e.what() that is being discarded with a generic catch. Is there interesting information to be scraped and reported from an e.what()?

Reply:
fair call. would like to extend the stream class to pass through the error callback such that subclasses can report proper errors instead of having to throw (filesystems shouldn't report these things as exceptions, should be part of the api)